go/src/koding/klient/uploader/uploader.go

```diff
@@ -0,0 +1,275 @@
+	// File path is also used as a S3 key.
+	File string `json:"file"`
+
+	// Interval makes the uploader stream the File at the given interval.
+	// By default File is not streamed.
+	//
+	// Ignored when File is empty and Content is used instead.
+	// Minimal interval is 15m.
+	Interval time.Duration `json:"interval"`
+
+	// Content is a one-time log content to upload, requires key to be non-empty.
+	// Superseded by File, if non-empty.
+	Content []byte `json:"content"`
+
+	// Key is required when Content is set.
+	Key string `json:"key"`
+}
+
+// Valid validates the request.
+func (req *UploadRequest) Valid() error {
+	if req.File == "" && len(req.Content) != 0 && req.Key == "" {
+		return errors.New("missing key")
+	}
+
+	if req.File != "" {
+		if _, err := os.Stat(req.File); err != nil {
+			return err
+		}
+
+		req.File = filepath.ToSlash(filepath.Clean(req.File))
+	}
+
+	return nil
+}
+
+func (req *UploadRequest) key() string {
+	if len(req.Content) != 0 && req.File == "" {
+		return req.Key
+	}
+
+	return req.File
+}
+
+// UploadResponse represents a response of the "log.upload" kite method.
+type UploadResponse struct {
+	URL string `json:"url"`
+}
+
+// UploadFile uploads the given file to the S3.
+//
+// If interval is > 0 then the given file is uploaded to S3 at the given interval.
+func (up *Uploader) UploadFile(file string, interval time.Duration) (string, error) {
+	req := &UploadRequest{
+		File:     file,
+		Interval: interval,
+	}
+
+	return up.upload(req)
+}
+
+// Upload is a kite handler for the "log.upload" method.
+func (up *Uploader) Upload(r *kite.Request) (interface{}, error) {
+	if r.Args == nil {
+		return nil, errors.New("missing argument")
+	}
+
+	var req UploadRequest
+	if err := r.Args.One().Unmarshal(&req); err != nil {
+		return nil, err
+	}
+
+	s, err := up.upload(&req)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UploadResponse{
+		URL: s,
+	}, nil
+}
+
+// Close stops uploading logs to the S3.
+func (up *Uploader) Close() error {
+	if up.close != nil {
+		close(up.close)
+		up.close = nil
+	}
+
+	return nil
+}
+
+func (up *Uploader) upload(req *UploadRequest) (string, error) {
+	if err := req.Valid(); err != nil {
+		return "", err
+	}
+
+	up.req <- req
+
+	return up.url(req), nil
+}
+
+func (up *Uploader) url(req *UploadRequest) string {
+	return fmt.Sprintf("https://%s.s3.amazonaws.com/%s", up.cfg.Bucket, path.Clean(up.cfg.Kite.Config.Id+"/"+req.key()))
+}
+
+func (up *Uploader) log() kite.Logger {
+	if up.cfg.Log != nil {
+		return up.cfg.Log
+	}
+
+	return defaultLog
+}
+
+func (up *Uploader) process() {
+	var (
+		files   = make(map[int]string)
+		tickers []*time.Ticker
+		cases   = []reflect.SelectCase{
+			0: {Dir: reflect.SelectRecv, Chan: reflect.ValueOf(up.close)},
+			1: {Dir: reflect.SelectRecv, Chan: reflect.ValueOf(up.req)},
+		}
+		prefix = up.cfg.Kite.Config.Id
+	)
+
+	for {
+		var req *UploadRequest
+
+		switch n, v, _ := reflect.Select(cases); n {
+		case 0:
```

Comment:
I think we have different takes on readability :) 

Since you said  "your call" lets change it to non-reflected version not to fiddle with the runtime memory 🙋